jax/_src/lax/parallel.py

```diff
@@ -500,26 +526,32 @@ def maybe_take(self, char: str, on_eof: bool = False):
   def parse_arg(self):
     subscripts = []
     names = []
+    # we continue parsing the argument until we hit eof or a ',' outside of an
+    # axis brace. In the latter case, we expect a subsequent argument.
     while not self.eof:
-      subscript, cont = self.parse_subscript()
-      if not cont: break
-      subscripts.append(subscript)
-    if self.eof:
-      return False, (subscripts, names)
-    if self.maybe_take(','):
-      return True, (subscripts, names)
-    else:
-      assert self.maybe_take('{')
-      first = True
-      while not self.maybe_take('}'):
-        if not first:
-          assert self.maybe_take(',')
-        first = False
-        if self.eof:
-          raise ValueError("Unterminated named axis brace")
-        axis_name = self.parse_axis_name()
-        names.append(axis_name)
-      return self.maybe_take(',', False), (subscripts, names)
+      # if we found a ',' outside of axis braces we're done.
+      # but expect a subsequent argument.
+      if self.maybe_take(','):
+        return True, (subscripts, names)
+      # a brace signifies the beginning of a sequence of named axes.
+      elif self.maybe_take('{'):
+        # continue parsing until the brace closes.
+        first = True
+        while not self.maybe_take('}'):
+          if not first:
+            assert self.maybe_take(',')
+          first = False
+          if self.eof:
+            raise ValueError("Unterminated named axis brace")
+          axis_name = self.parse_axis_name()
+          names.append(axis_name)
+      else:
+        # we're at a subscript
+        subscript = self.parse_subscript()
+        subscripts.append(subscript)
+
+    # we hit eof, so we don't expect a subsequent argument.
+    return False, (subscripts, names)
```

Comment:
Was the previous logic buggy somehow?

Reply:
I don't think it was buggy. It just didn't support subscripts and named axes in arbitrary order (just supported named axes at the end) and also it was not raising a ValueError for invalid characters (like having '0' as a subscript would raise the ValueError now)

Specifically, the logic before was:

1. keep consuming subscripts until either eof, an axis brace, a comma or an invalid character is hit.
2. if eof was hit, return the subscripts and (empty) named axes, and `cont = False` (which means no more arguments).
3. ...or if comma was hit, return the subscripts and (empty) named axes, and `cont = True`.
4. ...or if axis brace was hit, parse all the named axes inside `{...}`. And `cont = True/False` depending on whether the next character is a comma.

The change to the parser can be summarized as:
 (a) Expand the original while loop (where each iteration consumed a subscript) to either consume a subscript or a `{...}` block containing multiple axis names in each iteration. This means we now support subscripts, followed by groups of named axes interspersed in any order (so `ab{c,d}e{f,g}h` is also valid input.)
 (b) Remove the responsibility of detecting axis brace or comma from `parse_subscript` (the while loop body takes care of that) so that it can raise a ValueError if it doesn't see a lowercase ascii character.

The other important side-change I can think of is: because the earlier method checked axis braces last, it could do an assert if it failed to parse a subscript or detect a comma. (`assert self.maybe_take('{')`). But now it doesn't do that anymore, as it would go into the `parse_subscript... raise ValueError` code path, and raise an error because it's expecting a subscript.

Reply:
Is there any good argument for allowing named axes to be interspersed with subscripts?

Reply:
The main argument I have is ergonomics / readability. It's easier to switch to a version without named axes to one with _some_ named axes, but not imposing further constraints on the ordering of subscripts/named axes in the einsum spec. 

It can hurt readability if the user is forced to write `yz{x}` when the natural ordering is `{x}yz`. Of course, here the user might eventually use only named axes, i.e. `{x,y,z}` in which case there is no problem. But I think what might arise in practice is a transition period from no-named-axes to some-named-axes (maybe only eventually to all-named-axes). This is also because, currently, many ops already work with one 'batch' named axis, but very few work with all named axes.

That said, this is not a strong reason - so if you prefer the simpler API contract of named axes appearing together after all the subscripts, I can revert this part of the change.